File: apps/desktop/src/lib/ground.ts
/** The colour the next launch opens on.
 *
 *  A window is on screen for about a fifth of a second before the app's own stylesheet
 *  has been read, and in that time the page shows the webview's default - white - unless
 *  the page itself says otherwise before anything is fetched. index.html says it, off
 *  the value written down here: the ground this window had when it was last looked at.
 *
 *  Remembered rather than computed, because it cannot be computed from the markup. A
 *  theme file may set any colour it likes, and translucency sets none at all - the
 *  ground is transparent so the platform's material shows through, and a launch that
 *  painted a colour there would flash a solid window over somebody's wallpaper. So the
 *  one honest source is what the window actually resolved to, and that is what this
 *  reads.
 *
 *  Cosmetic by construction, and never load-bearing: a machine that has never run the
 *  app, or one whose storage was cleared, opens on the palette index.html carries. See
 *  the inline script there, and test/ground.test.ts, which holds the two to each other.
 */

import { keep } from './stored'
import { invoke, isDesktop } from './tauri'

/** Where it is kept. The same string is in index.html, which is read before any module
 *  of ours exists and so cannot import this; ground.test.ts holds the two together. */
const KEY = 'nib:ground'

/** Whether a read is already scheduled for the next frame. */
let asked = false

/** Writes down what the window is standing on, after the frame that changed it.
 *
 *  On the next frame rather than now, because every caller has just set an attribute on
 *  the root element and the computed value of a custom property is what it is after the
 *  style has been recalculated. Asked twice in one frame it reads once: the theme and
 *  the material can change together, and one of those changes is the ground. */
export function rememberGround(): void {
  if (asked || typeof requestAnimationFrame !== 'function') return
  asked = true

  requestAnimationFrame(() => {
    asked = false

    // The hint index.html painted, taken off the moment the real sheet is in place: it
    // is an inline style and would otherwise outrank every rule for the life of the
    // window, so a reader who changed the theme would keep the ground of the theme
    // before it. Taken off here rather than anywhere else because this is the first
    // code that runs with the app's own stylesheet applied - and because the value read
    // below has to be the sheet's, not the hint's.
    document.documentElement.style.removeProperty('background')

    // The resolved colour rather than the token: `--window-ground` is a custom property
    // and what it holds is whatever was written there, while a background colour read
    // off an element is always `rgb(…)` or `rgba(…)` - one shape, with the alpha in it,
    // which is what both readers of this need. A translucent window comes back as
    // `rgba(0, 0, 0, 0)`, and that is the answer, not a missing one.
    const ground = getComputedStyle(document.body).backgroundColor.trim()
    if (!ground) return

    keep(KEY, ground)

    // And the same value where the crate can reach it, because the crate cannot read
    // local storage and it is the crate that opens the window. Nothing waits for this
    // and nothing depends on it: a launch whose file is missing opens the way it always
    // did. See src-tauri/src/ground.rs.
    if (isDesktop) void invoke('remember_ground', { colour: ground }).catch(() => undefined)
  })
}

File: apps/desktop/src-tauri/src/ground.rs
//! The colour the window opens on, so that it can open before there is anything in
//! it.
//!
//! A window described in `tauri.conf.json` is built by the runtime before any of our
//! own code runs, and it is built around its webview: nothing is on screen until the
//! webview runtime has started. On Windows that was measured at 323 of the 778
//! milliseconds a launch of a five thousand note space takes - a third of the wait,
//! with the screen empty for all of it and this process using no processor at all.
//! So `ready` in lib.rs builds the window itself instead, and shows it at once.
//!
//! Which needs a colour, and the colour is the whole difficulty. A window has to open
//! on *this* reader's ground: the side of the theme they chose, a colour out of a
//! theme file they installed, or - with translucency on - no colour whatsoever, so
//! that the platform's own material shows through. None of that can be known before
//! the page that knows it has been read, so it is remembered instead: the window
//! writes down what it was standing on, and the next launch stands on that.
//!
//! Two readers of one value. This file is the crate's, for the native window; the
//! other is three lines inline in index.html, for the page - written from
//! `src/lib/ground.ts`, which keeps both. The page's copy is in local storage because
//! it is read before any module exists; this one is a file because the crate cannot
//! read local storage.
//!
//! Cosmetic by construction. A machine with nothing remembered - the first launch of
//! all - and a reader whose ground is the platform's material rather than a colour
//! both get the window exactly as it always came: built hidden, shown once there is
//! something in it. That is right rather than merely cautious. A solid rectangle in
//! the wrong colour, or an opaque one where somebody chose to see their wallpaper, is
//! worse than a moment more of nothing.

use std::path::PathBuf;
use tauri::utils::config::Color;
use tauri::AppHandle;

use crate::paths::{config_dir, made, write_atomically};

/// What the file is called. Beside the themes and the history rather than in the log
/// folder: it is something the app keeps about itself, not something it reports.
const FILE: &str = "ground.txt";

/// Anything longer than this is not a colour and is not read. A `rgba()` with four
/// numbers in it is under thirty characters.
const MOST: usize = 64;

fn file(app: &AppHandle) -> Result<PathBuf, String> {
    Ok(config_dir(app)?.join(FILE))
}

/// What the window is standing on, as the window itself resolved it.
///
/// One string, written whenever it changes, which is a theme being chosen or
/// translucency being switched. The value is whatever `getComputedStyle` gives for the
/// page's background - `rgb(14, 16, 19)`, or `rgba(0, 0, 0, 0)` for a window with the
/// material showing through - so no colour arithmetic happens on either side of this:
/// the browser resolves it and the crate reads back what the browser said.
#[tauri::command(async)]
pub fn remember_ground(app: AppHandle, colour: String) -> Result<(), String> {
    if colour.len() > MOST {
        return Err("that is not a colour".into());
    }

    let path = file(&app)?;
    if let Some(parent) = path.parent() {
        made(parent)?;
    }

    write_atomically(&path, colour.as_bytes())
}

/// The colour the next window may open on, or nothing.
///
/// Nothing for every case that is not a colour to paint: no file yet, a file that is
/// not one, and - deliberately - any colour that is not fully opaque. A ground with
/// alpha in it is a reader who asked to see the material behind their window, and
/// painting it on the window layer is exactly what would hide that.
pub fn remembered(app: &AppHandle) -> Option<Color> {
    let said = std::fs::read_to_string(file(app).ok()?).ok()?;
    colour_of(said.trim())
}

/// `rgb(14, 16, 19)` or `rgba(14, 16, 19, 1)`, as a browser writes a resolved colour.
///
/// Nothing for anything else, which includes `transparent`, `rgba(…, 0)` and every
/// half-written file: the caller's answer to nothing is the window it has always
/// built, so a value this cannot read costs a launch its head start and never
/// anything else.
fn colour_of(said: &str) -> Option<Color> {
    let inside = said
        .strip_prefix("rgba(")
        .or_else(|| said.strip_prefix("rgb("))?
        .strip_suffix(')')?;

    let mut parts = inside.split(',').map(str::trim);
    let red = parts.next()?.parse::<u8>().ok()?;
    let green = parts.next()?.parse::<u8>().ok()?;
    let blue = parts.next()?.parse::<u8>().ok()?;

    // The fourth is optional and, when it is there, has to be 1: a browser writes the
    // alpha as a decimal fraction, and anything under one is a window somebody wants to
    // see through.
    match parts.next() {
        None => {}
        Some(alpha) if alpha.parse::<f32>().is_ok_and(|one| one >= 1.0) => {}
        Some(_) => return None,
    }

    if parts.next().is_some() {
        return None;
    }

    Some(Color(red, green, blue, 255))
}

#[cfg(test)]
mod tests {
    use super::colour_of;
    use tauri::utils::config::Color;

    #[test]
    fn a_resolved_colour_is_the_colour_it_says() {
        assert_eq!(colour_of("rgb(14, 16, 19)"), Some(Color(14, 16, 19, 255)));
        assert_eq!(
            colour_of("rgb(251,252,253)"),
            Some(Color(251, 252, 253, 255))
        );
        assert_eq!(
            colour_of("rgba(14, 16, 19, 1)"),
            Some(Color(14, 16, 19, 255))
        );
    }

    /// The one case that matters most: a reader who chose to see the material behind
    /// their window must not get a solid rectangle painted over it for a third of a
    /// second. Both spellings a browser uses for it.
    #[test]
    fn a_ground_somebody_can_see_through_is_no_ground_to_paint() {
        assert_eq!(colour_of("rgba(0, 0, 0, 0)"), None);
        assert_eq!(colour_of("rgba(14, 16, 19, 0.5)"), None);
        assert_eq!(colour_of("transparent"), None);
    }

    /// And anything else at all, because the file is on a disk and disks keep halves
    /// of things. Every one of these is a launch that opens the way it always did.
    #[test]
    fn nothing_that_is_not_a_colour_is_read_as_one() {
        for said in [
            "",
            "rgb(",
            "rgb(14, 16)",
            "rgb(14, 16, 19",
            "rgb(14, 16, 19, 1, 1)",
            "rgb(300, 16, 19)",
            "rgb(-1, 16, 19)",
            "#0e1013",
            "var(--bg)",
            "rgb(a, b, c)",
        ] {
            assert_eq!(colour_of(said), None, "{said}");
        }
    }
}
